### ISSUE TITLE
Fails to generate schema for github-workflow

### ISSUE THREAD

**Issue by frezbo** [omissis/go-jsonschema#112]
Trying to generate schema for github workflow from https://github.com/SchemaStore/schemastore/blob/master/src/schemas/json/github-workflow.json

```bash
❯ gojsonschema -v -p main github-workflow.json
go-jsonschema: Loading github-workflow.json
go-jsonschema: Failed: error parsing from file github-workflow.json: error parsing JSON file github-workflow.json: failed to unmarshal JSON: failed to unmarshal schema: failed to unmarshal type: failed to unmarshal type: failed to unmarshal type: failed to unmarshal type: failed to unmarshal type: json: cannot unmarshal array into Go value of type schemas.ObjectAsType
```

**Comment by omissis** [comment]
Hi @frezbo , thanks for the heads up, I will take a look as soon as I can :)

**Comment by frezbo** [comment]
I had this diff trying to debug, didn't spent too much time, seems some fields are defined as list whereas the code expects strings

```diff
diff --git pkg/schemas/model.go pkg/schemas/model.go
index 82c0d24..4a986ac 100644
--- pkg/schemas/model.go
+++ pkg/schemas/model.go
@@ -26,6 +26,7 @@ package schemas
 import (
 	"encoding/json"
 	"fmt"
+	"os"
 )
 
 // Schema is the root schema.
@@ -40,7 +41,7 @@ type Schema struct {
 func (s *Schema) UnmarshalJSON(data []byte) error {
 	var unmarshSchema unmarshalerSchema
 	if err := json.Unmarshal(data, &unmarshSchema); err != nil {
-		return fmt.Errorf("failed to unmarshal schema: %w", err)
+		return fmt.Errorf("failed to unmarshal schema into base: %w", err)
 	}
 
 	// Fall back to id if $id is not present.
@@ -54,7 +55,7 @@ func (s *Schema) UnmarshalJSON(data []byte) error {
 	}
 
 	if err := json.Unmarshal(data, &legacySchema); err != nil {
-		return fmt.Errorf("failed to unmarshal schema: %w", err)
+		return fmt.Errorf("failed to unmarshal schema into legacy: %w", err)
 	}
 
 	if unmarshSchema.Definitions == nil && legacySchema.Definitions != nil {
@@ -156,6 +157,30 @@ type Type struct {
 	GoJSONSchemaExtension *GoJSONSchemaExtension `json:"goJSONSchema,omitempty"` //nolint:tagliatelle // breaking change
 }
 
+func lineAndCharacter(input string, offset int) (line int, character int, err error) {
+	lf := rune(0x0A)
+
+	if offset > len(input) || offset < 0 {
+		return 0, 0, fmt.Errorf("Couldn't find offset %d within the input.", offset)
+	}
+
+	// Humans tend to count from 1.
+	line = 1
+
+	for i, b := range input {
+		if b == lf {
+			line++
+			character = 0
+		}
+		character++
+		if i == offset {
+			break
+		}
+	}
+
+	return line, character, nil
+}
+
 // UnmarshalJSON accepts booleans as schemas where `true` is equivalent to `{}`
 // and `false` is equivalent to `{"not": {}}`.
 func (value *Type) UnmarshalJSON(raw []byte) error {
@@ -172,7 +197,20 @@ func (value *Type) UnmarshalJSON(raw []byte) error {
 
 	var obj ObjectAsType
 	if err := json.Unmarshal(raw, &obj); err != nil {
-		return fmt.Errorf("failed to unmarshal type: %w", err)
+		if jsonError, ok := err.(*json.UnmarshalTypeError); ok {
+			fmt.Println("field: ", jsonError.Field)
+			fmt.Println("struct: ", jsonError.Struct)
+			fmt.Println("type: ", jsonError.Type)
+			fmt.Println("value: ", jsonError.Value)
+			fmt.Println("offser: ", jsonError.Offset)
+			fmt.Println("raw: ", string(raw))
+			line, character, lcErr := lineAndCharacter(string(raw), int(jsonError.Offset))
+			fmt.Fprintf(os.Stderr, "test %d failed with error: Cannot parse JSON schema due to a syntax error at line %d, character %d: %v\n", 0, line, character, jsonError.Error())
+			if lcErr != nil {
+				fmt.Fprintf(os.Stderr, "Couldn't find the line and character position of the error due to error %v\n", lcErr)
+			}
+		}
+		return fmt.Errorf("failed to unmarshal type as onjectastype: %w", err)
 	}
 
 	// Take care of legacy fields from older RFC versions.
@@ -182,7 +220,7 @@ func (value *Type) UnmarshalJSON(raw []byte) error {
 		Definitions  Definitions      `json:"definitions,omitempty"` // Section 5.26.
 	}{}
 	if err := json.Unmarshal(raw, &legacyObj); err != nil {
-		return fmt.Errorf("failed to unmarshal type: %w", err)
+		return fmt.Errorf("failed to unmarshal type legacyObj: %w", err)
 	}
 
 	if legacyObj.Definitions != nil && obj.Definitions == nil {

```

**Comment by skandragon** [comment]
I have the same issue with parsing https://raw.githubusercontent.com/CycloneDX/specification/1.5/schema/bom-1.5.schema.json.  The issue is with the definition of `LicenseChoice`, where it is an array of `oneOf` structure choices, I believe.

**Comment by nmiculinic** [comment]
Still an issue

**Comment by fire833** [comment]
I managed to get things working by applying the following patch to the schema. As far as I know, no structural changes occur, it is just some validation that changes, and I personally don't really care about validation, I just want the models for free. Hope this helps:

```diff

--- workflows_old.json	2025-06-05 10:11:01.829177588 -0500
+++ workflows.json	2025-06-05 10:24:44.246026899 -0500
@@ -441,14 +441,7 @@
         "step": {
             "type": "object",
             "additionalProperties": false,
-            "dependencies": {
-                "working-directory": [
-                    "run"
-                ],
-                "shell": [
-                    "run"
-                ]
-            },
+            "dependencies": {},
             "oneOf": [
                 {
                     "required": [
@@ -754,11 +747,9 @@
                             "type": "array",
                             "anyOf": [
                                 {
-                                    "items": [
-                                        {
-                                            "type": "string"
-                                        }
-                                    ],
+                                    "items": {
+                                        "type": "string"
+                                    },
                                     "minItems": 1,
                                     "additionalItems": {
                                         "type": "string"


```

Update: well this kind of works, a lot of the nested types don't get rendered, but I'm not sure what the cause is for that. Apologies.

**Comment by ghostsquad** [comment]
I came here to say the same thing. Just tried:

```
go-jsonschema \
  --schema-package=https://json.schemastore.org/github-workflow.json=github.com/cakehappes/gocto \
   --schema-output=https://json.schemastore.org/github-workflow.json=schema1.go \
   github-workflow.json
go-jsonschema: Failed: error parsing from file github-workflow.json: cannot load schema
failed to load schema "github-workflow.json": error parsing JSON file github-workflow.json: failed to unmarshal JSON: failed to unmarshal schema: failed to unmarshal type: failed to unmarshal type: json: cannot unmarshal array into Go value of type schemas.ObjectAsType
```

for reference, I'm getting github-workflow.json from: https://github.com/SchemaStore/schemastore/blob/master/src/schemas/json/github-workflow.json

**Comment by jeremylong-snc** [comment]
Facing the same issue with the CycloneDx 1.6 schema:

```bash
curl https://raw.githubusercontent.com/CycloneDX/specification/refs/heads/master/schema/bom-1.6.schema.json --output pkg/schema/bom-1.6.schema.json
go-jsonschema -p pkg/models/cyclonedx pkg/schema/bom-1.6.schema.json 
```

Results in:
```
go-jsonschema: Failed: error parsing from file pkg/schema/bom-1.6.schema.json: cannot load schema
failed to load schema "pkg/schema/bom-1.6.schema.json": error parsing JSON file pkg/schema/bom-1.6.schema.json: failed to unmarshal JSON: failed to unmarshal schema: failed to unmarshal type: failed to unmarshal type: failed to unmarshal type: json: cannot unmarshal array into Go value of type schemas.ObjectAsType
```

**Comment by plheide** [comment]
Root cause is a single parser gap. In draft-07 `dependencies` is dual-form — each value is either a schema or an array of property names — but `pkg/schemas/model.go` models only the schema form. `github-workflow.json` uses the array form in exactly two places, both under `definitions/step` (`working-directory` and `shell`), and those fail `Type.UnmarshalJSON`. That is the `cannot unmarshal array into Go value of type schemas.ObjectAsType` error, and it is why @fire833's patch worked — replacing `dependencies` with `{}` removes the array form.

Worth noting for anyone arriving from the older comments: the tuple-form `items` that used to be the second blocker is no longer present in the current SchemaStore schema, so `dependencies` is now the only thing failing.

Opened #623 with a fix and a regression fixture. With it, `github-workflow.json` generates 44 types that compile.

One honest caveat, matching @fire833's follow-up about nested types: 19 declarations still come out as `interface{}`, notably
`type GithubWorkflowJsonJobs map[string]interface{}`. So this makes the schema generate rather than fully modelled — `jobs` in particular is still untyped. That is a composition limitation rather than a parsing one and probably deserves its own issue.

**Comment by plheide** [comment]
@jeremylong-snc — your CycloneDX case is a different construct from the one #623 fixes, so I want to correct any impression that #623 would unblock you.

`bom-1.6.schema.json` has no array-form `dependencies`. What it has is tuple-form `items` — an array of schemas rather than a single one — at `definitions/licenseChoice/oneOf/1`:

```json
{
  "type": "array",
  "additionalItems": false,
  "minItems": 1,
  "maxItems": 1,
  "items": [ { "type": "object", "required": ["expression"], ... } ]
}
```

`Items` is modelled as a single `*Type`, so the array fails the same way. This is the *other* blocker named in the earlier comments here — github-workflow dropped it, which is why #623 unblocks that schema but not yours.

I have opened #624 for it. With that, CycloneDX 1.6 generates 172 types that compile clean. A one-element tuple (your case, and the common "tuple of exactly one" idiom) maps onto that element's type, so you get a properly typed slice; a heterogeneous tuple warns and falls back to an untyped array rather than inventing positional semantics.

One caveat on scope: #624 unblocks *loading*. `licenseChoice` itself is a `oneOf` over two **array** variants ("Multiple licenses" and "SPDX License Expression"), and none of the existing composition strategies reached that shape — discriminated `oneOf` (#566) needs object variants sharing a required const property, the try-each fallback (#567) was object-only, and the primitive-union wrapper (#549/#574) handles scalars.

#567 to covers this. Array variants dispatch on the key-sets of their *elements*, which is what tells these two apart (`license` versus
`expression`). `licenseChoice` now generates as a dispatching holder rather than `[]interface{}`:

```go
type LicenseChoice struct {
	Variant0 *LicenseChoiceVariant0 `json:"-" yaml:"-"`
	Variant1 *LicenseChoiceVariant1 `json:"-" yaml:"-"`
}

// A list of SPDX licenses and/or named licenses.
type LicenseChoiceVariant0 []struct {
	License License `json:"license" ...`
}

// A tuple of exactly one SPDX License Expression.
type LicenseChoiceVariant1 []struct { ... }
```

`[{"license":...}]` selects Variant0, `[{"expression":...}]` selects Variant1, and input matching neither (or both) is a clear error rather than a silent `interface{}`.

Setting expectations honestly, though: #624 is small and standalone, whereas #567 sits on top of a series of stacked PRs, so it is not something you can pick up on its own in the near term. If you need CycloneDX generating today, #624 alone gets you a compiling 172-type package with `licenseChoice` left untyped.
